src/main/java/org/broadinstitute/hellbender/engine/MultiFeatureWalker.java

```diff
@@ -0,0 +1,264 @@
+     */
+    private void setDictionaryAndSamples() {
+        dictionary = getMasterSequenceDictionary();
+        if ( hasReference() ) {
+            dictionary = betterDictionary(reference.getSequenceDictionary(), dictionary);
+        }
+        if ( hasReads() ) {
+            dictionary = betterDictionary(reads.getSequenceDictionary(), dictionary);
+        }
+        for ( final FeatureInput<? extends Feature> input : features.getAllInputs() ) {
+            final Object header = features.getHeader(input);
+            if ( header instanceof SVFeaturesHeader ) {
+                final SVFeaturesHeader svFeaturesHeader = (SVFeaturesHeader)header;
+                dictionary = betterDictionary(svFeaturesHeader.getDictionary(), dictionary);
+                final List<String> sampleNames = svFeaturesHeader.getSampleNames();
+                if ( sampleNames != null ) {
+                    samples.addAll(svFeaturesHeader.getSampleNames());
+                }
+            } else if (header instanceof VCFHeader ) {
+                final VCFHeader vcfHeader = (VCFHeader)header;
+                dictionary = betterDictionary(vcfHeader.getSequenceDictionary(), dictionary);
+                samples.addAll(vcfHeader.getSampleNamesInOrder());
+            }
+        }
+        if ( dictionary == null ) {
+            throw new UserException("No dictionary found.  Provide one as --" +
+                    StandardArgumentDefinitions.SEQUENCE_DICTIONARY_NAME + " or --" +
+                    StandardArgumentDefinitions.REFERENCE_LONG_NAME + ".");
+        }
+    }
+
+    /**
+     * Makes sure that the two dictionaries are consistent with regard to contig names and order.
+     * Returns the more comprehensive (larger) dictionary if they're consistent.
+     */
+    private static SAMSequenceDictionary betterDictionary( final SAMSequenceDictionary newDict,
+                                                           final SAMSequenceDictionary curDict ) {
+        if ( curDict == null ) return newDict;
+        if ( newDict == null ) return curDict;
+        final SAMSequenceDictionary smallDict;
+        final SAMSequenceDictionary largeDict;
+        if ( newDict.size() <= curDict.size() ) {
+            smallDict = newDict;
+            largeDict = curDict;
+        } else {
+            smallDict = curDict;
+            largeDict = newDict;
+        }
+        int lastIdx = -1;
+        for ( final SAMSequenceRecord rec : smallDict.getSequences() ) {
+            final int newIdx = largeDict.getSequenceIndex(rec.getContig());
+            if ( newIdx == -1 ) {
+                throw new UserException("Contig " + rec.getContig() +
+                                        " not found in the larger dictionary");
+            }
+            if ( newIdx <= lastIdx ) {
+                throw new UserException("Contig " + rec.getContig() +
+                                        " not in same order as in larger dictionary");
```

Comment:
Since these are exceptions that actual users are likely to trigger, it would be helpful to include the contents of both dictionaries in the error messages. You can do this by throwing a `UserException.IncompatibleSequenceDictionaries` and passing the two dictionaries in.